tasks/centos.yml

```diff
@@ -5,6 +5,75 @@
     state: present
   become: true
 
+- name: centos | check for nvme devices
+  shell: |
+    set -o pipefail
+    cat /proc/partitions | awk '{print $4}' | grep -q nvme. ; echo $?
+  args:
+    executable: /bin/bash
+  register: blkdev_nvme
+  changed_when: false
+
+- name: centos | check for amazon ebs devices
+  shell: |
+    set -o pipefail
+    lsblk -O -J | grep -qi "amazon elastic block store" ; echo $?
+  register: blkdev_awsebs
+  changed_when: false
+  args:
+    executable: /bin/bash
+
+- name: centos | set flag for nvme subsystem
+  set_fact:
+    device_is_nvme: "{{ true if (blkdev_nvme.stdout == '0') else false }}"
+
+- name: centos | set flag for aws ebs devices
+  set_fact:
+    device_is_awsebs: "{{ true if (blkdev_awsebs.stdout == '0') else false }}"
+
+- name: Block to map renamed ec2 ebs nvmeXnY devices to their original sdX/xvdX names
+  when: device_is_nvme and device_is_awsebs|bool
+  become: true
+  block:
+    - name: centos | download nvme mapping binary helper
+      get_url:
+        url: "{{ ebsnvme_binary_helper_url }}"
+        dest: "{{ ebsnvme_binary_helper_tmp }}/"
+      changed_when: false
+
+    - name: centos | extract binary helper
+      unarchive:
+        src: "{{ ebsnvme_binary_helper_tmp }}/{{ ebsnvme_binary_helper_file }}"
+        dest: "{{ ebsnvme_binary_helper_tmp }}/"
+        remote_src: yes
+        mode: 0755
+      changed_when: false
+
+    - name: centos | copy binary helper
+      copy:
+        src: "{{ ebsnvme_binary_helper_tmp }}/go-ebsnvme"
+        dest: "{{ ebsnvme_binary_helper_path }}"
+        mode: a+x
+        remote_src: yes
+
+    - name: centos | template script helper
+      template:
+        src: 'ebsnvme-id.j2'
+        dest: "{{ ebsnvme_scrip_helper_path }}"
+        mode: a+x
+
+    - name: centos | template udev rule
+      template:
+        src: '70-ec2-nvme-devices.rules.j2'
+        dest: '/etc/udev/rules.d/70-ec2-nvme-devices.rules'
+        mode: 0755
+      register: udev_rule
+
+    - name: centos | reload and trigger udev rules
+      shell:
+        cmd: udevadm control --reload-rules && udevadm trigger
+      when: udev_rule.changed
```

Comment:
Given that this code is specific to AWS, could you move it to a separate tasks file, guarded by something like this:

```
- include_tasks: amazon.yml
  when: ansible_distribution == 'Amazon'
```

Reply:
Hi @markgoddard, in reality specific to AWS but also to Centos (for sure centos8) and nitro EC2 instances (using new nvme driver)... so many specs to manage...

But are you sure about  this ? Now it is transparent and specific to centos (for backword compatibility with debian - probably can be ported but not teste as now).

Also, ansible_distribution in you idea where coma from ? a static configuration variable or set_fact in centos.yml o in other file (e.g main.tf)?

Reply:
Also, the code snippet is inside a conditional block. **"All block of tasks along this code snippet"*** we'll be entirely execute only for Amazon ec2 with ebs storage device using nvme driver. Not sure if this issue could be on other "components combination" too !?

Reply:
I see, you are right - we could hit this on any distro. What we want to know is whether we are running in AWS. I believe this check will do that:

```
"ansible_system_vendor" == "Amazon EC2"
```

My concern is around performance at scale. Even skipped tasks can take a significant time to execute when there are many hosts in the inventory. Using include_tasks is dynamic, and if skipped, the included tasks will not incur any penalty.

Checkout my investigations here: https://github.com/stackhpc/ansible-scaling

Reply:
"Ansible scaling" seems an interesting work, I'll read more as soon as possible.

I believe that the compromise between performance and automatic decisions should be considered.

Tell me more about the check you suggested, e.g. where in your proposition in the code should be assigned "Amazon EC" value to "ansible_system_vendor" variable ? From the currents dynamic set_fact

```
- name: centos | check for nvme devices
  shell: |
    set -o pipefail
    cat /proc/partitions | awk '{print $4}' | grep -q nvme. ; echo $?
  args:
    executable: /bin/bash
  register: blkdev_nvme
  changed_when: false

- name: centos | check for amazon ebs devices
  shell: |
    set -o pipefail
    lsblk -O -J | grep -qi "amazon elastic block store" ; echo $?
  register: blkdev_awsebs
  changed_when: false
  args:
    executable: /bin/bash

- name: centos | set flag for nvme subsystem
  set_fact:
    device_is_nvme: "{{ true if (blkdev_nvme.stdout == '0') else false }}"

- name: centos | set flag for aws ebs devices
  set_fact:
    device_is_awsebs: "{{ true if (blkdev_awsebs.stdout == '0') else false }}"
```

Or from a static choise as host variables ?

```
ansible_system_vendor: 'Amazon EC2'
```

Reply:
Concerns about performance at scale mitigated in cb792e5

Reply:
Perfect, thanks

Reply:
Hi @markgoddard to be consistent I need to change in amazon.yml all:
```
"- name: centos | ..."
```
to 
```
"- name: amazon | ..."
```

or
```
"- name: centos | amazon | ..."
```

what do you prefer (which is your comment standard for this) ?

Reply:
Probably just amazon makes sense.

Reply:
Done. Can you merge this PR ?

Reply:
I'm afraid I don't have the power, @mrlesmithjr does